### PR TITLE
Support the catalog overlay

### PR DESCRIPTION
**Pull request by markccchiang**
This is for issue #383. The backend support of opening catalog files, parsing and filtering the VOTable data. A few things to be done in the future:

- Add image bounds and region settings in the data filter
- Find a good way to deal with very large files, for example, loading partial data at the first of opening files and then load more data if necessary
- Benchmark the performance

### CODE REVIEW

**Review by veggiesaurus — CHANGES_REQUESTED** [review]
some conflicts with the protobuf changes

Due to complexity of the PR (and the frontend PR), and the build changes required, we should merge them after the code freeze

File: Main.cc [hunk 1]
```diff
@@ -460,11 +498,15 @@ int main(int argc, const char* argv[]) {
 
         session_number = 0;
 
-        websocket_hub.onMessage(&OnMessage);
-        websocket_hub.onConnection(&OnConnect);
-        websocket_hub.onDisconnection(&OnDisconnect);
+        // Create a hub group and pass the PERMESSAGE_DEFLATE option
+        group = websocket_hub.createGroup<uWS::SERVER>(uWS::PERMESSAGE_DEFLATE);
```
Comment (veggiesaurus) [inline Main.cc]: I think you can do this without creating a group explicitly, by just passing these arguments to the hub constructor directly.

File: Session.cc [hunk 2]
```diff
@@ -1243,7 +1283,7 @@ void Session::SendEvent(CARTA::EventType event_type, uint32_t event_id, google::
     head->icd_version = carta::ICD_VERSION;
     head->request_id = event_id;
     message.SerializeToArray(msg.data() + sizeof(carta::EventHeader), message_length);
-    _out_msgs.push(msg);
+    _out_msgs.push(std::make_pair(msg, compress));
```
Comment (veggiesaurus) [inline Session.cc]: doesn't this result in an additional copy? you should rather form the pair first, and then serialize to the pair's `msg.data()`

File: Session.cc [hunk 3]
```diff
@@ -1257,10 +1297,14 @@ void Session::SendFileEvent(int32_t file_id, CARTA::EventType event_type, uint32
 void Session::SendPendingMessages() {
     // Do not parallelize: this must be done serially
     // due to the constraints of uWS.
-    std::vector<char> msg;
+    std::pair<std::vector<char>, bool> msg;
     if (_connected) {
         while (_out_msgs.try_pop(msg)) {
-            _socket->send(msg.data(), msg.size(), uWS::BINARY);
+            if (msg.second) { // Compress the message
```
Comment (veggiesaurus) [inline Session.cc]: this can be simplified to just `_socket->send(msg.first.data(), msg.first.size(), uWS::BINARY, nullptr, nullptr, msg.second);`

**Comment by markccchiang** [comment]
@veggiesaurus I have changed codes by your suggestions. But wondering is it the problem of the frontend's version, the image can not be shown on the viewer on my own test(?)

**Comment by veggiesaurus** [comment]
> @veggiesaurus I have changed codes by your suggestions. But wondering is it the problem of the frontend's version, the image can not be shown on the viewer on my own test(?)

I noticed that earlier on mix 'n match. Odd.

**Comment by markccchiang** [comment]
Thanks! With the new version of Qi's catalog branch, the image is back on the viewer. Probably the backend and frontend use different protobuf version.

**Pull request merged**